File: src/graphics/magma/lib/magma_service/sys_driver/performance_counters_server.cc
// Copyright 2023 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "performance_counters_server.h"

#include <lib/driver/logging/cpp/logger.h>

namespace msd::internal {
zx::result<> PerformanceCountersServer::Create(fdf::OutgoingDirectory* outgoing) {
  zx_status_t status = zx::event::create(0, &event_);
  if (status != ZX_OK) {
    return zx::error(status);
  }

  auto perf_counter_access_protocol =
      [this](fidl::ServerEnd<fuchsia_gpu_magma::PerformanceCounterAccess> server_end) mutable {
        fidl::BindServer(dispatcher_, std::move(server_end), this);
      };

  fuchsia_gpu_magma::PerformanceCounterService::InstanceHandler handler({
      .access = std::move(perf_counter_access_protocol),
  });

  {
    auto status = outgoing->template AddService<fuchsia_gpu_magma::PerformanceCounterService>(
        std::move(handler));
    if (status.is_error()) {
      FDF_LOG(ERROR, "%s(): Failed to add service to outgoing directory: %s\n", __func__,
              status.status_string());
      return status.take_error();
    }
  }

  return zx::ok();
}

zx_koid_t PerformanceCountersServer::GetEventKoid() {
  zx_info_handle_basic_t info;
  if (event_.get_info(ZX_INFO_HANDLE_BASIC, &info, sizeof(info), nullptr, nullptr) != ZX_OK)
    return 0;
  return info.koid;
}

void PerformanceCountersServer::GetPerformanceCountToken(
    GetPerformanceCountTokenCompleter::Sync& completer) {
  zx::event event_duplicate;
  zx_status_t status = event_.duplicate(ZX_RIGHT_SAME_RIGHTS, &event_duplicate);
  if (status != ZX_OK) {
    completer.Close(status);
  } else {
    completer.Reply(std::move(event_duplicate));
  }
}

}  // namespace msd::internal

File: src/graphics/magma/lib/magma_service/sys_driver_rust/src/performance_counters_server.rs
// Copyright 2026 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use anyhow::Context;
use fidl_fuchsia_gpu_magma as fidl_magma;
use futures::TryStreamExt;

pub struct PerformanceCountersServer {
    event: zx::Event,
}

impl PerformanceCountersServer {
    pub fn new() -> Result<Self, i32> {
        let event = zx::Event::create();
        Ok(PerformanceCountersServer { event })
    }

    pub fn get_event_koid(&self) -> Result<u64, i32> {
        self.event.koid().map(|koid| koid.raw_koid()).map_err(|status| status.into_raw())
    }

    pub async fn run(
        &self,
        mut stream: fidl_magma::PerformanceCounterAccessRequestStream,
    ) -> anyhow::Result<()> {
        while let Some(request) = stream.try_next().await.context("Stream error")? {
            match request {
                fidl_magma::PerformanceCounterAccessRequest::GetPerformanceCountToken {
                    responder,
                } => {
                    let duplicate_event = self
                        .event
                        .duplicate_handle(zx::Rights::SAME_RIGHTS)
                        .context("Duplicate handle failed")?;
                    responder.send(duplicate_event).context("Send failed")?;
                }
            }
        }
        Ok(())
    }
}
